### PR TITLE
Add provider_relationship_permissions table

### PR DESCRIPTION
**Pull request by steventux**
## Context

We're working out how to implement provider permissions at an organisational level. We'll need somewhere to store these permissions.

## Changes proposed in this pull request

Adds underlying `provider_relationship_permissions` table for provider-to-provider permissions.
Adds `view_safeguarding_information` permission as part of the table definition.

## Guidance to review

Naming things is hard, please suggest a good name for this permissions model. (`ProviderPermissions` is currently in use for user level permissions, we could change this though).

Some other names:

- `ProviderToProviderPermissions`
- `RelatedProviderPermissions`
- `ProviderOrganisationPermissions`

## Link to Trello card

https://trello.com/c/F6QCZXqx/2169-provider-to-provider-permissions-affect-what-users-can-do-in-the-system

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

**Comment by duncanjbrown** [comment]
I like the way you've prevented STI by naming the columns. 👨‍⚕️ 

I think the name ought to have "permissions" in it. `ProviderRelationshipPermissions`?

**Comment by steventux** [comment]
@duncanjbrown Ah! that wasn't intentional when naming the columns. This is very much a discussion PR. Happy to go with STI as we all agreed this was a good case for it.

**Comment by duncanjbrown** [comment]
Oh yeah, we still need STI don't we. Sorry — my bad

**Comment by steventux** [comment]
@duncanjbrown I've amended the table name to `provider_relationship_permissions` and the relationship columns to `provider_id` and `related_provider_id` so this should work with STI. There's also an added index for `type` + `provider_id`

File: db/migrate/20200519102317_add_provider_relationship_permissions.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    add_foreign_key :provider_relationship_permissions, :providers, column: :related_provider_id
+
+    add_index(
+      :provider_relationship_permissions,
+      %i[type provider_id],
+      unique: true,
+    )
+    add_index(
+      :provider_relationship_permissions,
+      %i[provider_id related_provider_id],
```
Comment (duncanjbrown) [inline db/migrate/20200519102317_add_provider_relationship_permissions.rb:23]: Wouldn't this cause a bit of trouble? If we had provider 1 ratifying courses at provider 2, then provider 2 would define two relationships: the "as ratifying provider" relationship and the "as training provider" relationship. We'd have to make sure we didn't put the same IDs in `provider_id`/`related_provider_id` in each case, which in practice would mean checking for an existing relationship then reversing the positions. 

Would it work to rename these columns as they were (`training_provider_id` and `ratifying_provider_id`) and making a single, three-column unique index of `ratifying_provider_id`, `training_provider_id` and `type`?

Comment (steventux) [reply db/migrate/20200519102317_add_provider_relationship_permissions.rb:23]: I would expect that to produce 2 records:

```
<Provider id: 1> # Ratifying provider
<Provider id: 2> # Training provider

# The STI relationships:
<AccreditedBodyPermissions provider_id: 1, related_provider_id: 2, a_permission: false>
<TrainingProviderPermissions provider_id: 2, related_provider_id: 1, a_permission: true>
```

If we condensed these into a single record with specific column names for ratifying and training providers would we then lose the granularity of the permissions on each relationship?

I like the compound index idea and I'm happy to make the column names more specific to the relationship, maybe I'm misunderstanding the use of the `type` here.

Comment (duncanjbrown) [reply db/migrate/20200519102317_add_provider_relationship_permissions.rb:23]: Ah, I think we agree. I'm proposing we make the role of each provider explicit in the column name. I might well be missing something, though

```
AccreditedBodyPermissions.create!(training_provider: 1, accredited_body: 2) => OK
TrainingProviderPermissions.create!(training_provider: 1, accredited_body: 2) => OK
AccreditedBodyPermissions.create!(training_provider: 1, accredited_body: 2) => 💥
```

**Pull request merged**